File: examples/05_encoder_led_animation.py
# import modules:
import board
import neopixel # neopixel.py should be in /lib
from encoder_simple import Encoder # encoder_simple.py should be in /lib

# define program constants
NUM_PIXELS = 9
COLOR_OFF = (0, 0 ,0)
COLOR_ON = (255, 0, 0)

# define program variables
pixel_on_index = 0

# create pixels object:
pixels = neopixel.NeoPixel(
    board.GP16,
    NUM_PIXELS,
    brightness=0.3,
    auto_write=False
)

# create our Encoder object:
encoder = Encoder(
    board.GP17, # CLK pin
    board.GP18, # DT pin
)

# reset the leds to their initial state:
pixels.fill((0, 0, 0))
pixels[pixel_on_index] = COLOR_ON
pixels.show()

while True:

    # get the latest rotation value:
    rotation = encoder.updateRotation()

    if rotation != None:
        # turn OFF the pixel that is currently on
        pixels[pixel_on_index] = COLOR_OFF

        # increment or decrement pixel_on_index:
        if rotation == "clockwise":
            pixel_on_index += 1

            # wrap around if pixel_on_index is greater than NUM_PIXELS:
            if pixel_on_index >= NUM_PIXELS:
                pixel_on_index = 0

        elif rotation == "couterclockwise":
            pixel_on_index -= 1
            
            # wrap around if pixel_on_index is smaller than 0:
            if pixel_on_index < 0:
                pixel_on_index = NUM_PIXELS - 1

        # turn ON the new pixel that is currently on
        pixels[pixel_on_index] = COLOR_O

        # update the physical led strip:
        pixels.show()

File: arduino/05_encoder_led_animation.cpp
#include <FastLED.h> // FastLED library should be installed
#include <RoxMux.h>  // RoxMux library should be installed

#define PIN 6         // pin where the data line of the LED strip is connected
#define NUM_PIXELS 10 // number of LEDs on the strip

#define PIN_CLK 4 // CLK pin for the encoder
#define PIN_DT 5  // DT pin for the encoder

CRGB pixels[NUM_PIXELS];          // our pixels array used to control the strip
int pixel_on_index = 0;           // index of the currently lit pixel
unsigned long last_timestamp = 0; // last update timestamp

RoxEncoder encoder;

void setup()
{
    // initialize the LED strip
    FastLED.addLeds<NEOPIXEL, PIN>(pixels, NUM_PIXELS);

    // set brightness (a bit lower than max to save our eyes)
    FastLED.setBrightness(100); // (min: 0, max: 255)

    // clear data and turn the entire strip off
    FastLED.clearData();
    fill_solid(pixels, NUM_PIXELS, CRGB::Black);
    pixels[pixel_on_index] = CRGB::Red;
    FastLED.show();

    // initialize the encoder pins
    pinMode(PIN_CLK, INPUT);
    pinMode(PIN_DT, INPUT);

    Serial.begin(9600);

    encoder.setTickSensitivity(1);
}

void loop()
{
    // read pin values to update encoder:
    encoder.update(digitalRead(PIN_CLK), digitalRead(PIN_DT), 5, LOW);

    // if encoder was rotated, update the LED strip:
    if (encoder.read())
    {
        // turn OFF the pixel that is currently on
        pixels[pixel_on_index] = CRGB::Black;

        // increment or decrement pixel_on_index:
        if (encoder.clockwise())
        {
            pixel_on_index += 1;

            // wrap around if pixel_on_index is greater than NUM_PIXELS:
            if (pixel_on_index >= NUM_PIXELS)
            {
                pixel_on_index = 0;
            }
        }
        else
        {
            pixel_on_index -= 1;

            // wrap around if pixel_on_index is smaller than 0:
            if (pixel_on_index < 0)
            {
                pixel_on_index = NUM_PIXELS - 1;
            }
        }

        // turn ON the new pixel that is currently on
        pixels[pixel_on_index] = CRGB::Red;

        // update the physical led strip:
        FastLED.show();
    }
}